File: 83. AverageOfLevelsInBinaryTree.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class AverageOfLevelsInBinaryTree {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    public static List<Double> averageOfLevels(TreeNode root) {
        List<Double> res = new ArrayList<>();
        if (root == null) return res;
        Deque<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        while (!q.isEmpty()) {
            int size = q.size();
            double sum = 0;
            for (int i = 0; i < size; i++) {
                TreeNode n = q.poll();
                sum += n.val;
                if (n.left != null) q.offer(n.left);
                if (n.right != null) q.offer(n.right);
            }
            res.add(sum / size);
        }
        return res;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);
        System.out.println(averageOfLevels(root)); // [3.0, 14.5, 11.0]
    }
}

File: 83. AverageOfLevelsInBinaryTree.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode *newNode(int val) {
    struct TreeNode *node = malloc(sizeof(*node));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

/* Prints averages of each level in Java List format, e.g. [3.0, 14.5, 11.0] */
void averageOfLevels(struct TreeNode *root) {
    if (!root) {
        puts("[]");
        return;
    }
    struct TreeNode *q[1024];
    int head = 0, tail = 0;
    q[tail++] = root;
    double results[1024];
    int numLevels = 0;
    while (head < tail) {
        int size = tail - head;
        long long sum = 0;
        for (int i = 0; i < size; ++i) {
            struct TreeNode *node = q[head++];
            sum += node->val;
            if (node->left)  q[tail++] = node->left;
            if (node->right) q[tail++] = node->right;
        }
        results[numLevels++] = (double)sum / size;
    }
    putchar('[');
    for (int i = 0; i < numLevels; ++i) {
        if (i) printf(", ");
        printf("%.1f", results[i]);
    }
    puts("]");
}

int main(void) {
    struct TreeNode *root = newNode(3);
    root->left  = newNode(9);
    root->right = newNode(20);
    root->right->left  = newNode(15);
    root->right->right = newNode(7);
    averageOfLevels(root); // [3.0, 14.5, 11.0]
    return 0;
}
